### PR TITLE
Update proteinortho to 6.0.6

### PR DESCRIPTION
**Pull request by BiocondaBot**
Update [`proteinortho`](https://bioconda.github.io/recipes/proteinortho/README.html): **6.0.5** &rarr; **6.0.6**

[install with bioconda](http://bioconda.github.io/recipes/proteinortho/README.html) [Conda](https://anaconda.org/bioconda/proteinortho/files)

Info | Link
-----|-----
Recipe | [`recipes/proteinortho`](https://github.com//bioconda/bioconda-recipes/tree/bump/proteinortho/recipes/proteinortho) (click to view/edit other files)
Releases |[https://gitlab.com/paulklemm_PHD/proteinortho/tags](https://gitlab.com/paulklemm_PHD/proteinortho/tags)

***

This pull request was automatically generated (see [docs](https://bioconda.github.io/updating.html)).

### CODE REVIEW

**Comment by pmjklemm** [comment]
Not sure what failed, but it works on [jenkins](https://jenkins.brew.sh/job/Homebrew%20Core%20Pull%20Requests/44892/) (mac) and with [gitlabci](https://gitlab.com/paulklemm_PHD/proteinortho/pipelines/70551948) (linux)
```
TESTS FAILED: proteinortho-6.0.6-h9428810_0.tar.bz2
COMMAND FAILED: docker run -t --net host --rm -v /tmp/tmpu9c9jhz7/build_script.bash:/opt/build_script.bash -v /home/circleci/project/miniconda/conda-bld:/opt/host-conda-bld -v /home/circleci/project/recipes/proteinortho:/opt/recipe -e LANG=en_US.UTF-8 -e HOST_USER_ID=1001 bioconda/bioconda-utils-build-env:latest /bin/bash /opt/build_script.bash
```
any idea what is going wrong here?

**Comment by npavlovikj** [comment]
The error is:
```
18:48:00 BIOCONDA INFO (OUT) /opt/conda/conda-bld/proteinortho_1562870630975/test_tmp/run_test.sh: /opt/conda/conda-bld/proteinortho_1562870630975/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placeho/bin/proteinortho: /usr/bin/perl: bad interpreter: No such file or directory
```

I wonder if `perl` needs to be added in the `host` section as well.

**Comment by pmjklemm** [comment]
You patched the main file but i have a lot of perl files in src/. Is there not a better way? Also i have to replace the interpreter everytime I update from now?

**Comment by npavlovikj** [comment]
Ah @pmjklemm, I added changes to the branch that just passed :/ Can you please revert yours?

**Comment by npavlovikj** [comment]
I tried adding `perl` in both the `host` and `run` sections, but the error was still there. Next, I patched `proteinortho6.pl` and replaced `#!/usr/bin/perl` with `#!/opt/anaconda1anaconda2anaconda3/bin/perl`. 
This seemed to work and the tests passed.

`#!/usr/bin/perl` used to work too, but because I don't know if that not working now is intentional or bug, I am pinging @bioconda/core so they can give us some insight.

**Comment by npavlovikj** [comment]
> `/opt/anaconda1anaconda2anaconda3/bin/perl` seems silly

Actually, `/opt/anaconda1anaconda2anaconda3/bin/perl` is not silly and is the right placeholder to use in the recipes (you can search for that placeholder here https://readthedocs.com/projects/continuumio-conda/downloads/pdf/latest/). 
`/opt/anaconda1anaconda2anaconda3` guarantees that the correct innterpreter is always used. On the other hand, `/usr/bin/perl` is poining to a different perl, that may or may not exist. 
You can try using `/usr/bin/env perl` to be more general with your scripts, but even that may not always work depending on what is in `$PATH`.

**Comment by pmjklemm** [comment]
Sorry for undoing your patch, it works but i am not happy with it.
i was using `#!/usr/bin/env perl` until the PPA (apt-get) forced me into changing it to `#!/usr/bin/perl`. Would be unneccessary work to deploy different versions for conda with a different perl interpreter.

**Comment by npavlovikj** [comment]
> Sorry for undoing your patch, it works but i am not happy how it is.
> i was using `#!/usr/bin/env perl` until the PPA (apt-get) forced me into changing it to `#!/usr/bin/perl`. Would be unneccessary work to deploy different versions for conda with a different perl interpreter.

Ah, the switch from `#!/usr/bin/env perl` to `#!/usr/bin/perl` explains the error then! I don't know how many perl scripts there are in the package, but would it be that hard to create patch with the change when you create the recipe? You can have a bash script that iterates over all .pl files and replaces the shebang with the correct one. There are plenty recipes here that are patched that way, and you won't need to deploy different versions for conda and apt-get...

**Comment by pmjklemm** [comment]
Ahh ok, then i will stick to the patch solution for now. Currently i fiddle around with the patch syntax.
Thank you for your time (:

**Comment by mbargull** [comment]
> Actually, `/opt/anaconda1anaconda2anaconda3/bin/perl` is not silly and is the right placeholder to use in the recipes (you can search for that placeholder here https://readthedocs.com/projects/continuumio-conda/downloads/pdf/latest/).

Yes and no ;)

TL;DR : You don't have to (IMO shouldn't) use `/opt/anaconda1anaconda2anaconda3` directly but can rather let `conda-build` handle it for you. Just make sure the shebangs use the `${PREFIX}` path, e.g. `#! /opt/conda/conda-bld/.../_h_env_placehold.../perl`.

I know that the placeholder is documented and a few recipes on Bioconda use it directly. However, I would rather treat it as an implementation detail, if possible. It is just some "arbitrary" placeholder used by `conda` to know where to place the environment's prefix/path when a package is installed. On the build side, `conda-build` will prepare a package by automatically replacing occurrences of the expanded `${PREFIX}` variable (e.g., `/opt/conda/conda-bld/.../_h_env_placehold...`) with that placeholder, without the need for you to do it manually.
You can also see it as a means of communication between the package builder (i.e., `conda-build`) and package manager (i.e., `conda`). And that communication can (IMHO should) be seen as "private" in the sense that the users of both programs, `conda` as well as `conda-build`, needn't use it directly.

In the best-case scenario, the source's build, e.g., via `./configure`, `make` or similar, already does that for you (either automatically or with you providing some `build_var="${PREFIX}/bin/perl"` or something). If this is not the case, you can't use static `.patch` files, unfortunately, since we want to replace with the `${PREFIX}` variable's content. Meaning, you have to resort to some `sed` scripts or similar or dynamically create patch files and manually run `patch`.

Regarding `/usr/bin/env some-prog`:
Yes, you could use that and `conda` will even replace the placeholder with that if the install prefix is too long for a shebang line. But that can fail sometimes, e.g.,
- when (for whatever reason) the Conda environment is not the first entry with a `some-prog` executable on `${PATH}`, or
- when you start with `#! .../some-prog --some-arg` which is converted to `#! /usr/bin/env some-prog --some-param` but your shebang interpreter might only allow at most one parameter and interprets it as `/usr/bin/env 'some-prog --some-param`.
  Shebang are "fun", meaning non-standardized, fragile and with many limitations -- they were just not designed for how we use them today, e.g., with `conda`. Highly recommended read: https://www.in-ulm.de/~mascheck/various/shebang/

**Comment by pmjklemm** [comment]
Very interesting @mbargull but I still dont get it fully. 

1. So the env solution is fine ?! but your first critique point concerns me since I cannot ensure that the order of path is correct (although i dont know why this is a problem if perl comes before conda in path for the execution of a perl script). Also stated in the ulm link:
> In practice, env should not be a script.

2. So makes a lot of sense to treat these placeholder-variables „private“. But i dont have a configure file, so i could include this in my makefile: i just replace the perl shebang with 

> #! /opt/conda/conda-bld/.../_h_env_placehold.../perl

and then i should be golden? Except the conda-build replaces before making...

3. if 2. is also no solution, then i would use the patch system to replace all /usr/bin/perl with  /opt/conda/conda-bld/.../_h_env_placehold.../perl

Or any other idea to tell the meta.yml to use perl from path?
